src/main/groovy/com/google/protobuf/gradle/plugins/ProtobufPlugin.groovy

```diff
@@ -0,0 +1,86 @@
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ * list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ *
+ * 3. Neither the name of the copyright holder nor the names of its contributors
+ * may be used to endorse or promote products derived from this software without
+ * specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package com.google.protobuf.gradle.plugins
+
+import org.gradle.api.Action
+import org.gradle.api.GradleException
+import org.gradle.api.Plugin
+import org.gradle.api.Project
+import org.gradle.api.plugins.AppliedPlugin
+
+class ProtobufPlugin implements Plugin<Project> {
+    // any one of these plugins should be sufficient to proceed with applying this plugin
+    private static final List<String> javaPluginOptions = ['java']
+    private static final List<String> androidPluginOptions = ['com.android.application',
+                                                              'com.android.library',
+                                                              'android',
+                                                              'android-library']
+
+    private List<String> pluginsApplied = [];
+    private Project project;
+
+    void apply(final Project project) {
+        this.project = project
+
+        // At least one of the prerequisite plugins must by applied before this plugin can be applied, so
+        // we will use the PluginManager.withPlugin() callback mechanism to delay applying this plugin until
+        // after that has been achieved. If project evaluation completes before one of the prerequisite plugins
+        // has been applied then we will assume that none of prerequisite plugins were specified and we will
+        // throw an Exception to alert the user of this configuration issue.
+        applyWithPrerequisitePlugin(javaPluginOptions, 'com.google.protobuf.java')
```

Comment:
Yes, my first implementation is having separate tasks for C# and Java, to avoid mixing of options and a bit less coupled together. I have added [`ofCSharpSourceSet()`](https://github.com/ngyukman/protobuf-gradle-plugin/blob/csharp/src/main/groovy/com/google/protobuf/gradle/ProtobufConfigurator.groovy#L159)  to cater such case 

Indeed I didn't know `protoc` could be running for multiple target in one execution, but are we having performance problem on executing the same protobuf file twice yet? For building multiple language I think we can accept the extra time required for now. Moreover, after checking further a bit [`grpc`](http://www.grpc.io/docs/tutorials/basic/csharp.html#generating-client-and-server-code) could be required to apply separately for different tasks, having different task looks to be more flexible and easier to implement, and also keeping it similar to android, so easier for others to understand